docs/design/datacontracts/SList.md

```diff
@@ -0,0 +1,78 @@
+# Contract SList
+
+This contract allows reading and iterating over an SList data structure.
+
+## Data structures defined by contract
+``` csharp
+class SListReader
+{
```

Comment:
I think SList is unnecessary ceremony as a contract. Can we just have a next pointer in the Thread and a global that points to the first thread?

Reply:
I disagree. The idea of having a concept of how a linked list works that is external to the definition of any particular type works with it is quite useful. I don't expect that consumers of the cDAC will want to use the `SList` algorithm, but several algorithms within the cDAC may want to do so. For instance, the api for understanding what threads are blocked on a Monitor would use the `SList` algorithm as well on CoreCLR. Also as an example of how to handle our templated data structures, I believe it is fairly useful. We'll need to do something similar for `SHash`, and that will be a much more complex data structure to model in the contracts.

Although, looking a bit more closely at this particular use case, we might actually want to consider a doubly linked list instead of a singly linked list. The current model has a problem that cleaning up N threads has a O(N^2) cost instead of an O(N) cost since this is singly linked.

Finally, for V1 I'm describing the current the CoreCLR implementation of `SList` here right now, but honestly, the NativeAOT `SList` actually looks like a more reasonable implementation. This would be a fine place for us to wander by, and replace the CoreCLR `SList` with the NativeAOT `SList`, move one version number around, and see that we were able to take advantage of the granular versioning that the algorithms system provides.

Reply:
My experience has been that any abstractions for linked list that are more than a trivial helper function or a macro are useless.

I agree with you that an abstraction like this would be useful for hashtables.